torch/distributed/fsdp/flat_param.py

```diff
@@ -1517,12 +1530,24 @@ def _use_sharded_flat_param(self) -> None:
             )
         flat_param.data = flat_param._local_shard  # type: ignore[attr-defined]
         if self._use_orig_params:
-            self._use_sharded_views()
+            in_forward = self._training_state == HandleTrainingState.FORWARD
+            if (
+                in_forward
+                and self._sharding_strategy
+                in NO_RESHARD_AFTER_FORWARD_HANDLE_STRATEGIES
+            ):
+                self._skipped_use_sharded_views = True
+            else:
+                self._use_sharded_views()
             # For the post-forward reshard, we may try to use sharded gradient
             # views (or unsharded gradient views if a gradient was accumulated
             # in `no_sync()`), but for the post-backward reshard, we delay the
             # call to after the reduce-scatter.
-            if self._training_state == HandleTrainingState.FORWARD:
+            if (
+                in_forward
+                and self._sharding_strategy
+                not in NO_RESHARD_AFTER_FORWARD_HANDLE_STRATEGIES
```

Comment:
so we also skip calling use sharded grad views in this PR?

Reply:
Yes, I think we have to unless we want to use the `.data` hack to bypass the shape check since otherwise the parameters are unsharded while the gradients are sharded. This is a niche use case anyway since a gradient must be accumulated (either actually accumulated or `zero_grad(set_to_none=False)`).